src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataResourceSerializer.cs

```diff
@@ -634,6 +638,109 @@ public virtual ODataResource CreateResource(SelectExpandNode selectExpandNode, R
+
+            if(instanceAnnotationsDictionary != null)
+            {
+                if (instanceAnnotationsDictionary.ContainsKey(string.Empty))
+                {
+                    if (resource.InstanceAnnotations == null)
+                    {
+                        resource.InstanceAnnotations = new List<ODataInstanceAnnotation>();
+                    }
+
+                    foreach (KeyValuePair<string,object> annotation in instanceAnnotationsDictionary[string.Empty])
+                    {
+                        IEdmTypeReference edmTypeReference = resourceContext.SerializerContext.GetEdmType(annotation.Value,
+                           annotation.Value.GetType());
+                         
+                        ODataEdmTypeSerializer edmTypeSerializer = SerializerProvider.GetEdmTypeSerializer(edmTypeReference);
+
+                        if (edmTypeSerializer != null)
+                        {
+                            ODataValue annotationValue = edmTypeSerializer.CreateODataValue(annotation.Value, edmTypeReference, resourceContext.SerializerContext);
+
+                            if (annotationValue != null)
+                            {
+                                resource.InstanceAnnotations.Add(new ODataInstanceAnnotation(annotation.Key, annotationValue));
+                            }
+                        }
+                    }
+                }
+                
+                foreach(ODataProperty property in resource.Properties)
+                {
+                    string propertyName = property.Name;
+
+                    if (instanceAnnotationsDictionary.ContainsKey(propertyName))
+                    {
+                        if (property.InstanceAnnotations == null)
+                        {
+                            property.InstanceAnnotations = new List<ODataInstanceAnnotation>();
+                        }
+
+                        foreach (KeyValuePair<string,object> annotation in instanceAnnotationsDictionary[propertyName])
+                        {
+                            IEdmTypeReference edmTypeReference = resourceContext.SerializerContext.GetEdmType(annotation.Value,
+                            annotation.Value.GetType());
+
+                            ODataEdmTypeSerializer edmTypeSerializer = GetEdmTypeSerializer(edmTypeReference);
+                            if (edmTypeSerializer != null)
+                            {
+                                ODataValue annotationValue = edmTypeSerializer.CreateODataValue(annotation.Value, edmTypeReference, resourceContext.SerializerContext);
+
+                                if (annotationValue != null)
+                                {
+                                    property.InstanceAnnotations.Add(new ODataInstanceAnnotation(annotation.Key, annotationValue));
+                                }
+                            }
+                        }
+                    }
+                } 
```

Comment:
Code looks like a copy and paste of above - maybe extract to a new method with signature:

```
private void AddInstanceAnnotation(ResourceContext resourceContext, IDictionary<string, IDic...> instanceAnnotationsDictionary, string propertyName, ICollection<ODataInstanceAnnotation> edmInstanceAnnotations)
```

Reply:
@sreejith -- this is a good suggestion. Please move the common code from setting annotations on an instance and setting annotations on a property to a common routine for maintainability.

---
In reply to: [473392627](https://github.com/OData/WebApi/pull/2219#discussion_r473392627) [](ancestors = 473392627)